relays/messages/src/message_race_strategy.rs

```diff
@@ -107,47 +107,26 @@ where
 		// 2) we can't deliver new nonce until header, that has emitted this nonce, is finalized
 		// by target client
 		// 3) selector is used for more complicated logic
+		//
+		// => let's first select range of entries inside deque that are already finalized at
+		// the target client and pass this range to the selector
 		let best_header_at_target = race_state.best_finalized_source_header_id_at_best_target.clone()?;
-		let mut nonces_end = None;
+		self.source_queue
+			.iter()
+			.enumerate()
+			.take_while(|(_, (queued_at, _))| queued_at.0 <= best_header_at_target.0)
+			.map(|(index, _)| index)
+			.last()
+	}
+
+	/// Remove all nonces that are less than or equal to given nonce from the source queue.
+	pub fn remove_le_nonces_from_source_queue(&mut self, nonce: MessageNonce) {
 		while let Some((queued_at, queued_range)) = self.source_queue.pop_front() {
-			// select (sub) range to deliver
-			let queued_range_begin = queued_range.begin();
-			let queued_range_end = queued_range.end();
-			let range_to_requeue = if queued_at.0 > best_header_at_target.0 {
-				// if header that has queued the range is not yet finalized at bridged chain,
-				// we can't prove anything
-				Some(queued_range)
-			} else {
-				// selector returns `Some(range)` if this `range` needs to be requeued
-				selector(queued_range).await
-			};
-
-			// requeue (sub) range and update range to deliver
-			match range_to_requeue {
-				Some(range_to_requeue) => {
-					assert!(
-						range_to_requeue.begin() <= range_to_requeue.end()
-							&& range_to_requeue.begin() >= queued_range_begin
-							&& range_to_requeue.end() == queued_range_end,
-						"Incorrect implementation of internal `selector` function. Expected original\
-						range {:?} to end with returned range {:?}",
-						queued_range_begin..=queued_range_end,
-						range_to_requeue,
-					);
-
-					if range_to_requeue.begin() != queued_range_begin {
-						nonces_end = Some(range_to_requeue.begin() - 1);
-					}
-					self.source_queue.push_front((queued_at, range_to_requeue));
-					break;
-				}
-				None => {
-					nonces_end = Some(queued_range_end);
-				}
+			if let Some(range_to_requeue) = queued_range.greater_than(nonce) {
+				self.source_queue.push_front((queued_at, range_to_requeue));
```

Comment:
Just to confirm - `greater_than` does not return empty ranges, right?

Reply:
Yes, it returns `None` if there are nonces > given nonce (i.e. if range is empty).